sample_qc/joint_sample_qc.py

```diff
@@ -179,69 +168,57 @@ def tie_breaker(l, r):
-
-    key_expr = ['{0}_median'.format(metric) for metric in qc_metrics] + ['{0}_mad'.format(metric) for metric in qc_metrics]
-    value_expr = [hl.median(hl.agg.collect(ht['{}'.format(metric)])) for metric in qc_metrics] + \
-                 [1.4826 * hl.median(hl.abs(hl.agg.collect(ht['{0}'.format(metric)]) -
-                                            hl.median(hl.agg.collect(ht['{}'.format(metric)])))) for metric in qc_metrics]
-    ht = ht.group_by(ht.qc_pop, ht.qc_platform).aggregate(**dict(zip(key_expr, value_expr)))
-
-    upper_key_expr = ['{0}_upper'.format(metric) for metric in qc_metrics]
-    upper_value_expr = [ht['{0}_median'.format(metric)] + 4 * ht['{0}_mad'.format(metric)] if metric != 'callrate' else 1 for
-                        metric in qc_metrics]
-
-    lower_key_expr = ['{0}_lower'.format(metric) for metric in qc_metrics]
-    lower_value_expr = [ht['{0}_median'.format(metric)] - 4 * ht['{0}_mad'.format(metric)] if metric != 'callrate' else 0.99 for
-                        metric in qc_metrics]
-
-    annotation = dict(zip(upper_key_expr, upper_value_expr))
-    annotation.update(dict(zip(lower_key_expr, lower_value_expr)))
-    ht = ht.annotate(**annotation)
-    ht = ht.annotate(idx=ht.qc_pop + "_" + hl.str(ht.qc_platform))
-    ht = ht.filter(hl.is_defined(ht.idx))
-    ht.export(qc_temp_data_prefix(data_type) + '.platform_pop_MAD.txt.bgz')
-    return ht
-# TODO: investigate why some idx categories are NA (i.e., why some pop/platform information is missing)
-
-
-def apply_pop_platform_filters(ht: hl.Table, mt: hl.MatrixTable, qc_metrics: List[str]) -> hl.MatrixTable:
-    """
-    Flag samples that fall outside pop- and platform-specific QC metric thresholds
 
-    :param Table ht: Table containing upper- and lower-threshold values for each pop- and platform-specific cohort of samples
-    :param MatrixTable mt: MT containing samples to which pop- and platform-specific outlier filtering is applied
-    :param list qc_metrics: list of metrics on which to flag outlier samples
-    :return: MatrixTable containing the pop_platform_filter annotation summarizing metrics on which a given sample is an outlier
-    :rtype: MatrixTable
-    """
-    thresh_colnames = ['idx'] + ['{}_upper'.format(metric) for metric in qc_metrics] + [
-        '{}_lower'.format(metric) for metric in qc_metrics]
-    ht = ht.select(*thresh_colnames).key_by('idx')
-    ht.write('/tmp.h', True)  # temporary write/read for hail assert bug
-    ht = hl.read_table('/tmp.h')
-    mt = mt.annotate_cols(_thresholds=ht[mt.idx])
+    def make_pop_filters_expr(ht: hl.Table, qc_metrics: List[str]) -> hl.expr.SetExpression:
+        return hl.set(hl.filter(lambda x: hl.is_defined(x),
+                                [hl.or_missing(ht[f'fail_{metric}'], metric) for metric in qc_metrics]))
+
+    ht = ht.select(*strata, **ht.sample_qc.select(*qc_metrics)).key_by('s').persist()
+
+    def get_metric_expr(ht, metric):
+        return hl.bind(
```

Comment:
Do you still need `bind`s here? I thought it was deprecated with Hail's built-in CSE

Reply:
Yeah, that's not needed anymore. I think it was when I initially wrote that code. Anyhow, I'll refactor.